internal/controller/postgrescluster/postgres.go

```diff
@@ -602,3 +591,103 @@ func (r *Reconciler) reconcilePostgresWALVolume(
 
 	return pvc, err
 }
+
+// reconcileDatabaseInitSQL runs custom SQL files in the database. When
+// DatabaseInitSQL is defined, the function will find the primary pod and run
+// SQL from the defined ConfigMap
+func (r *Reconciler) reconcileDatabaseInitSQL(ctx context.Context,
+	cluster *v1beta1.PostgresCluster, instances *observedInstances) error {
+	log := logging.FromContext(ctx)
+
+	// Spec is not defined, unset status and return
+	if cluster.Spec.DatabaseInitSQL == nil {
+		if cluster.Status.DatabaseInitSQL != nil {
+			cluster.Status.DatabaseInitSQL = nil
+		}
+		return nil
+	}
+
+	// Spec is defined but status is already set, return
+	if cluster.Status.DatabaseInitSQL != nil {
+		return nil
+	}
+
+	// Based on the previous checks, the user wants to run sql in the database.
+	// Check the provided ConfigMap name and field path to ensure the a string
+	// exists in the ConfigMap data
+	var (
+		err  error
+		data string
+		sql  bytes.Buffer
+	)
+
+	getDataFromFieldRef := func() (error, string) {
+		cm := &corev1.ConfigMap{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      cluster.Spec.DatabaseInitSQL.Name,
+				Namespace: cluster.Namespace,
+			},
+		}
+		err := r.Client.Get(ctx, client.ObjectKeyFromObject(cm), cm)
+		if err != nil {
+			return err, ""
+		}
+
+		path := cluster.Spec.DatabaseInitSQL.FieldPath
+		if _, ok := cm.Data[path]; !ok {
+			log.Info("ConfigMap did not contain expected data at fieldPath", "ConfigMap", cm.Name, "FieldPath", path)
+			return nil, ""
+		}
+
+		return nil, cm.Data[path]
+	}
+
+	if err, data = getDataFromFieldRef(); err != nil {
+		return err
+	} else if data == "" {
+		return nil
+	}
+	_, _ = sql.WriteString(data)
+
+	// Now that we have the data provided by the user. We can check for a
+	// writable pod and get the podExecutor for the pod's database container
+	var podExecutor postgres.Executor
+	if pod, _ := instances.writablePod(naming.ContainerDatabase); pod != nil {
+		podExecutor = func(
+			_ context.Context, stdin io.Reader, stdout, stderr io.Writer, command ...string,
+		) error {
+			return r.PodExec(pod.Namespace, pod.Name, naming.ContainerDatabase, stdin, stdout, stderr, command...)
+		}
+	}
+	if podExecutor == nil {
+		log.Info("podExecutor is nil. Is there a pod with a writable database container?")
```

Comment:
:thinking: I'm wondering when "podExecutor" might be nil. Didn't we just check that the pod returned was writable?

Reply:
hmm, yeah we are defining `podExecutor` so it should never be nil - assuming pod isn't nil and we check that. Think the same applies earlier in the file too